File: failure-agent.py
#!/usr/bin/env python3
"""
failure-agent.py — reads a Playwright JSON report and explains each test failure
in plain English using Claude.

Usage:
    python failure-agent.py [path/to/results.json]

Default report path: test-results/results.json

Requires:
    pip install anthropic
    ANTHROPIC_API_KEY set in the environment
"""

import argparse
import json
import sys
from pathlib import Path

import anthropic

SYSTEM_PROMPT = """You are a QA engineer who explains Playwright test failures in simple, plain English.

For each failure you receive:
1. State the root cause in one clear sentence.
2. Explain what the test was trying to do and exactly why it broke.
3. Suggest one or two concrete fixes the developer can try.

Be concise and practical. Avoid jargon. The reader wants to understand and fix the issue quickly."""


def collect_failures(suites: list, parent_title: str = "") -> list[dict]:
    """Recursively walk the Playwright JSON suite tree and collect failed tests."""
    failures = []
    for suite in suites:
        title = f"{parent_title} > {suite['title']}".lstrip(" > ")

        if suite.get("suites"):
            failures.extend(collect_failures(suite["suites"], title))

        for spec in suite.get("specs", []):
            for test in spec.get("tests", []):
                if test.get("status") == "unexpected":
                    for result in test.get("results", []):
                        if result.get("status") in ("failed", "timedOut"):
                            failures.append(
                                {
                                    "suite": title,
                                    "spec": spec["title"],
                                    "project": test.get("projectName", ""),
                                    "status": result.get("status", "failed"),
                                    "duration_ms": result.get("duration", 0),
                                    "retry": result.get("retry", 0),
                                    "error": result.get("error") or {},
                                    "stdout": result.get("stdout") or [],
                                    "stderr": result.get("stderr") or [],
                                }
                            )
    return failures


def build_failure_context(failure: dict) -> str:
    """Format a failure dict into a prompt-friendly block of text."""
    parts = [
        f"Test suite : {failure['suite']}",
        f"Test name  : {failure['spec']}",
        f"Browser    : {failure['project'] or 'unknown'}",
        f"Status     : {failure['status']}",
        f"Duration   : {failure['duration_ms']} ms",
    ]
    if failure["retry"]:
        parts.append(f"Retry      : #{failure['retry']}")

    error = failure["error"]
    if error.get("message"):
        parts.append(f"\nError message:\n{error['message']}")

    if error.get("stack"):
        stack_lines = error["stack"].splitlines()[:30]
        parts.append("Stack trace (first 30 lines):\n" + "\n".join(stack_lines))

    stdout_text = "".join(
        chunk.get("text", "")
        for chunk in failure["stdout"]
        if isinstance(chunk, dict)
    )
    if stdout_text.strip():
        parts.append(f"\nTest stdout (truncated):\n{stdout_text[:500]}")

    return "\n".join(parts)


def explain_failure(client: anthropic.Anthropic, failure: dict, index: int, total: int) -> None:
    """Stream Claude's explanation for a single test failure to stdout."""
    header = f"Failure {index}/{total}: {failure['suite']} > {failure['spec']}"
    if failure["project"]:
        header += f"  [{failure['project']}]"

    print("=" * 70)
    print(header)
    print("=" * 70)

    context = build_failure_context(failure)

    # The system prompt is identical for every failure in this run, so
    # cache_control on the system block lets Claude reuse the cached prefix
    # after the first request — saving tokens across multiple failures.
    with client.messages.stream(
        model="claude-opus-4-7",
        max_tokens=1024,
        thinking={"type": "adaptive"},
        system=[
            {
                "type": "text",
                "text": SYSTEM_PROMPT,
                "cache_control": {"type": "ephemeral"},
            }
        ],
        messages=[
            {
                "role": "user",
                "content": (
                    "Explain this Playwright test failure in plain English:\n\n"
                    + context
                ),
            }
        ],
    ) as stream:
        for delta in stream.text_stream:
            print(delta, end="", flush=True)

    print("\n")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Explain Playwright test failures using Claude."
    )
    parser.add_argument(
        "report",
        nargs="?",
        default="test-results/results.json",
        help="Path to the Playwright JSON report (default: test-results/results.json)",
    )
    args = parser.parse_args()

    report_path = Path(args.report)
    if not report_path.exists():
        print(f"Error: report not found at '{report_path}'", file=sys.stderr)
        print(
            "Run Playwright with the JSON reporter, e.g.:\n"
            "  npx playwright test --reporter=json\n"
            "Or add the JSON reporter to playwright.config.ts.",
            file=sys.stderr,
        )
        sys.exit(1)

    try:
        report = json.loads(report_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        print(f"Error: could not parse JSON report: {exc}", file=sys.stderr)
        sys.exit(1)

    failures = collect_failures(report.get("suites", []))
    if not failures:
        stats = report.get("stats", {})
        print(
            f"No failures found in the report. "
            f"({stats.get('expected', 0)} passed, "
            f"{stats.get('skipped', 0)} skipped)"
        )
        return

    client = anthropic.Anthropic()

    print(f"\nFound {len(failures)} failure(s). Asking Claude to explain each one...\n")

    for i, failure in enumerate(failures, 1):
        explain_failure(client, failure, i, len(failures))


if __name__ == "__main__":
    main()

File: failure-agent.ts
#!/usr/bin/env npx ts-node
/**
 * failure-agent.ts — reads a Playwright JSON or JUnit XML report and explains
 * each test failure in plain English using Claude.
 *
 * Usage:
 *   npx ts-node failure-agent.ts [path/to/report]
 *
 * Supported formats:
 *   Playwright JSON  test-results/results.json  (default)
 *   JUnit XML        test-results/junit.xml      (fallback / explicit)
 *
 * Requires:
 *   ANTHROPIC_API_KEY set in the environment
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import Anthropic from '@anthropic-ai/sdk';
import { parseStringPromise } from 'xml2js';

const SYSTEM_PROMPT = `You are a QA engineer who explains Playwright test failures in simple, plain English.

For each failure you receive:
1. State the root cause in one clear sentence.
2. Explain what the test was trying to do and exactly why it broke.
3. Suggest one or two concrete fixes the developer can try.

Be concise and practical. Avoid jargon. The reader wants to understand and fix the issue quickly.`;

interface Failure {
  suite: string;
  name: string;
  project: string;
  durationMs: number;
  message: string;
  stack: string;
  stdout: string;
  stderr: string;
  retry: number;
}

// ── Playwright JSON ──────────────────────────────────────────────────────────

interface PwError { message?: string; stack?: string; }
interface PwResult {
  status: string;
  duration: number;
  retry: number;
  error?: PwError;
  stdout?: { text: string }[];
  stderr?: { text: string }[];
}
interface PwTest { projectName?: string; status: string; results: PwResult[]; }
interface PwSpec { title: string; ok: boolean; tests: PwTest[]; }
interface PwSuite { title: string; suites?: PwSuite[]; specs?: PwSpec[]; }
interface PwReport { suites: PwSuite[]; }

function collectFromSuite(suite: PwSuite, parentTitle: string): Failure[] {
  const title = parentTitle ? `${parentTitle} > ${suite.title}` : suite.title;
  const out: Failure[] = [];

  for (const spec of suite.specs ?? []) {
    if (spec.ok) continue;
    for (const test of spec.tests) {
      for (const result of test.results) {
        if (result.status !== 'failed' && result.status !== 'timedOut') continue;
        out.push({
          suite: title,
          name: spec.title,
          project: test.projectName ?? '',
          durationMs: result.duration,
          message: result.error?.message ?? '',
          stack: result.error?.stack ?? '',
          stdout: (result.stdout ?? []).map(l => l.text).join('').slice(0, 500),
          stderr: (result.stderr ?? []).map(l => l.text).join('').slice(0, 500),
          retry: result.retry,
        });
      }
    }
  }

  for (const child of suite.suites ?? []) {
    out.push(...collectFromSuite(child, title));
  }

  return out;
}

function parseJsonReport(filePath: string): Failure[] {
  const report: PwReport = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
  return report.suites.flatMap(s => collectFromSuite(s, ''));
}

// ── JUnit XML ────────────────────────────────────────────────────────────────

async function parseJunitReport(filePath: string): Promise<Failure[]> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const doc: any = await parseStringPromise(fs.readFileSync(filePath, 'utf-8'), {
    explicitArray: true,
  });

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const suites: any[] = doc.testsuites?.testsuite ?? (doc.testsuite ? [doc.testsuite] : []);
  const out: Failure[] = [];

  for (const suite of suites) {
    const suiteName: string = suite.$?.name ?? '';
    for (const tc of suite.testcase ?? []) {
      if (!tc.failure) continue;
      const f = tc.failure[0];
      out.push({
        suite: suiteName,
        name: tc.$?.name ?? '',
        project: '',
        durationMs: Math.round(parseFloat(tc.$?.time ?? '0') * 1000),
        message: f.$?.message ?? '',
        stack: (typeof f === 'string' ? f : (f._ ?? '')).trim(),
        stdout: (tc['system-out']?.[0] ?? '').slice(0, 500),
        stderr: (tc['system-err']?.[0] ?? '').slice(0, 500),
        retry: 0,
      });
    }
  }

  return out;
}

// ── Formatting ───────────────────────────────────────────────────────────────

function formatFailure(f: Failure): string {
  const parts = [
    `Test suite : ${f.suite}`,
    `Test name  : ${f.name}`,
  ];
  if (f.project) parts.push(`Project    : ${f.project}`);
  parts.push(`Duration   : ${(f.durationMs / 1000).toFixed(2)} s`);
  if (f.retry > 0) parts.push(`Retry      : ${f.retry}`);
  if (f.message) parts.push(`\nError message:\n${f.message}`);
  if (f.stack) {
    parts.push(`Stack trace (first 30 lines):\n${f.stack.split('\n').slice(0, 30).join('\n')}`);
  }
  if (f.stdout) parts.push(`\nTest stdout (truncated):\n${f.stdout}`);
  if (f.stderr) parts.push(`\nTest stderr (truncated):\n${f.stderr}`);
  return parts.join('\n');
}

// ── Claude streaming ─────────────────────────────────────────────────────────

async function explainFailure(
  client: Anthropic,
  failure: Failure,
  index: number,
  total: number,
): Promise<void> {
  const divider = '='.repeat(70);
  console.log(divider);
  console.log(`Failure ${index}/${total}: ${failure.suite} > ${failure.name}`);
  console.log(divider);

  // The system prompt is identical for every call in this run, so caching it
  // saves tokens on every request after the first.
  const stream = await client.messages.create({
    model: 'claude-opus-4-7',
    max_tokens: 1024,
    system: [
      {
        type: 'text',
        text: SYSTEM_PROMPT,
        cache_control: { type: 'ephemeral' },
      } as Anthropic.TextBlockParam & { cache_control: { type: 'ephemeral' } },
    ],
    messages: [
      {
        role: 'user',
        content: `Explain this Playwright test failure in plain English:\n\n${formatFailure(failure)}`,
      },
    ],
    stream: true,
  });

  for await (const event of stream) {
    if (
      event.type === 'content_block_delta' &&
      event.delta.type === 'text_delta'
    ) {
      process.stdout.write(event.delta.text);
    }
  }

  console.log('\n');
}

// ── Main ─────────────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  const DEFAULT_JSON = 'test-results/results.json';
  const DEFAULT_JUNIT = 'test-results/junit.xml';

  let reportPath = process.argv[2];

  if (!reportPath) {
    if (fs.existsSync(DEFAULT_JSON)) reportPath = DEFAULT_JSON;
    else if (fs.existsSync(DEFAULT_JUNIT)) reportPath = DEFAULT_JUNIT;
    else {
      console.error(
        `Error: no report found. Tried:\n  ${DEFAULT_JSON}\n  ${DEFAULT_JUNIT}\n\n` +
        'Run Playwright with a reporter, e.g.:\n' +
        "  npx playwright test --reporter=json\n" +
        'Or configure in playwright.config.ts:\n' +
        "  ['json', { outputFile: 'test-results/results.json' }]",
      );
      process.exit(1);
    }
  }

  if (!fs.existsSync(reportPath)) {
    console.error(`Error: report not found at '${reportPath}'`);
    process.exit(1);
  }

  const ext = path.extname(reportPath).toLowerCase();
  let failures: Failure[];

  if (ext === '.json') {
    failures = parseJsonReport(reportPath);
  } else if (ext === '.xml') {
    failures = await parseJunitReport(reportPath);
  } else {
    console.error(`Error: unsupported format '${ext}'. Pass a .json or .xml report file.`);
    process.exit(1);
  }

  if (failures.length === 0) {
    console.log('No failures found in the report.');
    return;
  }

  const client = new Anthropic();
  console.log(`\nFound ${failures.length} failure(s). Asking Claude to explain each one...\n`);

  for (let i = 0; i < failures.length; i++) {
    await explainFailure(client, failures[i], i + 1, failures.length);
  }
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});
